// react-native-intl-phone-input/src/IntlPhoneInput.js
import React from 'react';
import {
  View,
  Text,
  Modal,
  FlatList,
  StyleSheet,
  SafeAreaView,
  TouchableWithoutFeedback,
  TouchableOpacity,
  TextInput,
  Image, Keyboard
} from 'react-native';
import PropTypes from 'prop-types';
import data from './Countries';

export default class IntlPhoneInput extends React.Component {
  constructor(props) {
    super(props);
    const defaultCountry = data.filter((obj) => obj.code === props.defaultCountry)[0] || data.filter((obj) => obj.code === 'TR')[0];
    this.state = {
      defaultCountry,
      flag: defaultCountry.flag,
      modalVisible: false,
      dialCode: defaultCountry.dialCode,
      phoneNumber: '',
      mask: props.mask || defaultCountry.mask,
      countryData: data,
      selectedCountry: defaultCountry,
      placeholderTextColor: 'grey',
      onblur: false
    };
  }

  onChangePropText = (unmaskedPhoneNumber, phoneNumber) => {
    const { dialCode, mask, selectedCountry } = this.state;
    const countOfNumber = mask.match(/9/g).length;
    if (this.props.onChangeText) {
      const isVerified = countOfNumber === unmaskedPhoneNumber?.length && phoneNumber?.length > 0;
      this.props.onChangeText({
        dialCode, unmaskedPhoneNumber, phoneNumber, isVerified, selectedCountry
      });
    }
  }

  onChangeText = (value) => {
    let unmaskedPhoneNumber = (value.match(/\d+/g) || []).join('');

    if (unmaskedPhoneNumber.length === 0) {
      this.setState({ phoneNumber: '' });
      this.onChangePropText('', '');
      return;
    }


    let phoneNumber = this.state.mask.replace(/9/g, '_');
    for (let index = 0; index < unmaskedPhoneNumber.length; index += 1) {
      phoneNumber = phoneNumber.replace('_', unmaskedPhoneNumber[index]);
    }
    let numberPointer = 0;
    for (let index = phoneNumber.length; index > 0; index -= 1) {
      if (phoneNumber[index] !== ' ' && !isNaN(phoneNumber[index])) {
        numberPointer = index;
        break;
      }
    }
    phoneNumber = phoneNumber.slice(0, numberPointer + 1);
    unmaskedPhoneNumber = (phoneNumber.match(/\d+/g) || []).join('');

    this.onChangePropText(unmaskedPhoneNumber, phoneNumber);
    this.setState({ phoneNumber });
  }


  showModal = () => (this.props.disableCountryChange ? null : this.setState({ modalVisible: true }));

  hideModal = () => this.setState({ modalVisible: false });

  onCountryChange = async (code) => {
    const countryData = await data;
    try {
      const country = await countryData.filter((obj) => obj.code === code)[0];
      this.setState({
        dialCode: country.dialCode,
        flag: country.flag,
        mask: this.props.mask || country.mask,
        phoneNumber: '',
        selectedCountry: country,

      });
      this.hideModal();
    } catch (err) {
      const defaultCountry = this.state.defaultCountry;
      this.setState({
        dialCode: defaultCountry.dialCode,
        flag: defaultCountry.flag,
        mask: this.props.mask || defaultCountry.mask,
        phoneNumber: '',
        selectedCountry: defaultCountry,

      });
    }
  }

  filterCountries = (value) => {
    const { lang
    } = this.props;
    const countryData = data.filter((obj) => (obj[lang?.toLowerCase() ?? "en"]?.indexOf(value) > -1 || obj.dialCode.indexOf(value) > -1));
    this.setState({ countryData });
  }

  focus() {
    this.props.inputRef.current.focus();
  }

  renderModal = () => {
    if (this.props.customModal) return this.props.customModal(this.state.modalVisible, this.state.countryData, this.onCountryChange);
    const {
      countryModalStyle,
      modalContainer,
      modalFlagStyle,
      filterInputStyle,
      modalCountryItemCountryNameStyle,
      modalCountryItemCountryDialCodeStyle,
      closeText,
      filterText,
      searchIconStyle,
      closeButtonStyle,
      lang,
      placeholderTextColor
    } = this.props;
    return (
      <Modal transparent={true} visible={this.state.modalVisible}>

        <View style={[styles.modalContainer, modalContainer]}>
          <View style={{
            width: '100%',

            alignItems: 'center',
            backgroundColor: 'white', borderTopRightRadius: 20, borderTopLeftRadius: 20,
          }}>
            <View style={{ backgroundColor: 'lightgray', width: 70, height: 4, borderRadius: 10, margin: 5, }}></View>
            <View style={styles.filterInputStyleContainer}>

              <TextInput autoCapitalize="words" onChangeText={this.filterCountries} placeholder={filterText || 'Search here...'} style={[styles.filterInputStyle, filterInputStyle]} placeholderTextColor={placeholderTextColor} />
              <TouchableOpacity onPress={() => Keyboard.dismiss()}>
                <Image source={{ uri: 'https://cdn4.iconfinder.com/data/icons/standard-free-icons/139/Find01-512.png' }} style={{ height: 25, width: 25, paddingHorizontal: 5 }} />

              </TouchableOpacity>
            </View>
          </View>
          <FlatList
            style={{ flex: 1, backgroundColor: 'white' }}
            data={this.state.countryData}
            keyExtractor={(item, index) => index.toString()}
            renderItem={
              ({ item }) => (
                <TouchableOpacity onPress={() => this.onCountryChange(item.code)}>
                  <View style={[styles.countryModalStyle, countryModalStyle]}>
                    <Text style={[styles.modalFlagStyle, modalFlagStyle]}>{item.flag}</Text>
                    <View style={styles.modalCountryItemContainer}>
                      <Text style={[styles.modalCountryItemCountryNameStyle, modalCountryItemCountryNameStyle]}>{item[lang?.toLowerCase() ?? "en"]}</Text>
                      <Text style={[styles.modalCountryItemCountryDialCodeStyle, modalCountryItemCountryDialCodeStyle]}>{`  ${item.dialCode}`}</Text>
                    </View>
                  </View>
                </TouchableOpacity>
              )
            }
          />
        </View>
        {/* <TouchableOpacity onPress={() => this.hideModal()} style={[styles.closeButtonStyle, closeButtonStyle]}>
          <Text style={styles.closeTextStyle}>{closeText || 'CLOSE'}</Text>
        </TouchableOpacity> */}

      </Modal>
    );
  }

  renderAction = () => {
    const renderAction = this.props.renderAction;
    if (renderAction) {
      console.log("action", renderAction);
      if (typeof renderAction !== "function") throw ("The renderAction is not a function. Please set a renderAction function on there");
      else return this.props.renderAction();
    }
    return null;
  }

  render() {
    const { flag } = this.state;
    const {
      containerStyle,
      flagStyle,
      phoneInputStyle,
      dialCodeTextStyle,
      inputProps,
      placeholderTextColor
    } = this.props;
    return (
      <View style={{ ...styles.container, ...containerStyle }}>
        <TouchableOpacity onPress={() => this.showModal()}>
          <View style={styles.openDialogView}>
            <Text style={[styles.flagStyle, flagStyle]}>{flag}</Text>
            <Text style={[styles.dialCodeTextStyle, dialCodeTextStyle]}>{this.state.dialCode}</Text>
          </View>
        </TouchableOpacity>
        {this.renderModal()}
        <TextInput
          onBlur={Keyboard.dismiss}
          {...inputProps}
          style={[styles.phoneInputStyle, phoneInputStyle]}
          placeholder={'0000000000'}
          autoCorrect={false}
          keyboardType="number-pad"
          secureTextEntry={false}
          value={this.state.phoneNumber}
          onChangeText={this.onChangeText}
          placeholderTextColor={placeholderTextColor}

        />
        {this.renderAction()}

      </View>


    );
  }
}

IntlPhoneInput.propTypes = {
  lang: PropTypes.string,
  defaultCountry: PropTypes.string,
  mask: PropTypes.string,
  onChangeText: PropTypes.func,
  customModal: PropTypes.func,
  phoneInputStyle: PropTypes.object, // {}
  containerStyle: PropTypes.object, // {}
  dialCodeTextStyle: PropTypes.object, // {}
  flagStyle: PropTypes.object, // {}
  modalContainer: PropTypes.object, // {}
  filterInputStyle: PropTypes.object, // {}
  closeButtonStyle: PropTypes.object, // {}
  modalCountryItemCountryNameStyle: PropTypes.object, // {}
  filterText: PropTypes.string,
  closeText: PropTypes.string,
  searchIconStyle: PropTypes.object,
  disableCountryChange: PropTypes.bool,
  inputRef: PropTypes.object,
  placeholderTextColor: PropTypes.string
};

const styles = StyleSheet.create({
  closeTextStyle: {
    padding: 5,
    fontSize: 20,
    color: 'black',
    fontWeight: 'bold'
  },
  modalCountryItemCountryDialCodeStyle: {

    fontSize: 16, color: 'black', fontWeight: '700'
  },
  modalCountryItemCountryNameStyle: {
    flex: 1,
    fontSize: 16, color: 'black', fontWeight: '700'
  },
  modalCountryItemContainer: {
    flex: 1,
    paddingLeft: 15,
    flexDirection: 'row'
  },
  modalFlagStyle: {
    fontSize: 35,
  },
  modalContainer: {
    paddingTop: '100%',

    flex: 1,
    backgroundColor: 'rgba(0,0,0,0.5)'
  },
  flagStyle: {
    fontSize: 30,
  },
  dialCodeTextStyle: {
    fontSize: 16, color: 'black', fontWeight: '600'
  },
  countryModalStyle: {
    flex: 1,
    borderColor: 'black',
    padding: 8,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    paddingHorizontal: 30,
  },
  openDialogView: {
    flexDirection: 'row',
    alignItems: 'center',
    // borderWidth:1//flag view outside
  },
  filterInputStyle: {
    flex: 1,
    paddingTop: 10,
    paddingBottom: 10,
    backgroundColor: '#f1f1f1', borderRadius: 10, paddingLeft: 15,
    color: 'black',
    fontSize: 15, fontWeight: '700', marginRight: 10, height: 45
    // borderWidth: 1
  },
  searchIcon: {
    padding: 10,
  },
  filterInputStyleContainer: {
    width: '100%',
    flexDirection: 'row',
    alignItems: 'center',
    backgroundColor: 'white', borderTopRightRadius: 20, borderTopLeftRadius: 20, paddingHorizontal: 10,
  },
  phoneInputStyle: {
    fontSize: 16, fontWeight: '600', marginLeft: 10,
    flex: 1
  },
  container: {
    flexDirection: 'row',
    paddingHorizontal: 12,

    borderRadius: 10,
    alignItems: 'center',
    backgroundColor: 'white',
    marginHorizontal: 0, borderWidth: 1, height: 45, borderColor: '#9a9a9a'
  },
  searchIconStyle: {
    color: 'black',
    fontSize: 15,
    marginLeft: 15
  },
  buttonStyle: {
    alignItems: 'center',
    padding: 14,
    marginBottom: 10,
    borderRadius: 3,
  },
  buttonText: {
    fontSize: 18,
    fontWeight: 'bold',
    color: 'black',
  },
  countryStyle: {
    flex: 1,
    borderColor: 'black',
    borderTopWidth: 1,
    padding: 12,
  },
  closeButtonStyle: {

    alignItems: 'center',
    alignSelf: 'center'
  }
});
